crates/core/tedge_agent/src/tedge_operation_converter/actor.rs

```diff
@@ -278,6 +286,80 @@ impl TedgeOperationConverterActor {
                 log_file.log_script_output(&output).await;
                 Ok(())
             }
+            OperationAction::Command(sub_operation, input_excerpt, handlers) => {
+                let next_state = &handlers.on_exec.status;
+                info!(
+                    "Triggering {sub_operation} command, and moving {operation} operation to {next_state} state"
+                );
+
+                // Create the sub-command init state with a reference to its invoking command
+                let sub_cmd_id = self.id_generator.new_id();
```

Comment:
> Why not reuse the invoking command id itself so that it is easier to connect/correlate the entire command chain?

That's true that we can avoid to generate a truly new command id.

However, we cannot reuse directly the same command id, because this might confuse the c8y mapper, as it uses the prefix of the command ids to determine if it owns a command instance.

So, I will remove the use an id generator and simply prepend  an agent specific prefix to the cmd id generated by the mapper.

> Of course it can be done using the combination of the `__invoking_command__` and `__sub_command__` values in the payloads as well. But, I feel that a common id in the topic itself would have been just easier compared to those values in cluttered payloads across multiple messages, unless there's no other issue with that.

Reusing for the invoked command the command id of the invoking command is not enough. We also need the operation types. The invoked command must be able to resume the invoking command. And the invoking command needs to retrieve the final state of the invoked command.

Reply:
Thinking about these extra `__invoking_command__` and `__sub_command__` fields injected into the command and sub-command payloads. There is a way to get rid of them. Instead of storing the invocation relationships in the command payload, these relationships can be persisted in `/etc/tedge/.agent/workflows` along the current command states, without polluting the state payload. I will give a try to this idea.